src/cluster_legacy.c

```diff
@@ -143,8 +150,20 @@ static inline int getNodeDefaultReplicationPort(clusterNode *n) {
     return server.tls_replication ? n->tls_port : n->tcp_port;
 }
 
-int clusterNodeClientPort(clusterNode *n, int use_tls) {
-    return use_tls ? n->tls_port : n->tcp_port;
+int clusterNodeClientPort(clusterNode *n, int use_tls, client *c) {
+    int port = n->tcp_port;
+    int tlsport = n->tls_port;
+
+    if (c != NULL) {
+        if (n->announce_client_tcp_port) {
+            port = n->announce_client_tcp_port;
+        }
+        if (n->announce_client_tls_port) {
+            tlsport = n->announce_client_tls_port;
+        }
+    }
+
+    return use_tls ? tlsport : port;
```

Comment:
Nit. Here,_ `port` and `tlsport` are both assigned but only one is returned in the end. It looks suboptimal, but maybe it's optimized by the compiler.

Just thinking, can we refactor it in some way to make it more strait-forward. Is this equivalent?

```suggestion
int clusterNodeClientPort(clusterNode *n, int use_tls, client *c) {
    if (use_tls && c != NULL && n->announce_client_tls_port) {
        return n->announce_client_tls_port;
    } else if (use_tls) {
        return n->tls_port;
    } else if (c != NULL && n->announce_client_tcp_port) {
        return n->announce_client_tcp_port;
    } else {
        return n->tcp_port;
    }
```